datajunction-server/datajunction_server/sql/decompose.py

```diff
@@ -0,0 +1,146 @@
+
+    def __init__(self):
+        """Register handlers for aggregation functions"""
+        self.handlers = {
+            dj_functions.Sum: self._generic_additive_agg,
+            dj_functions.Count: self._generic_additive_agg,
+            dj_functions.Max: self._generic_additive_agg,
+            dj_functions.Min: self._generic_additive_agg,
+            dj_functions.Avg: self._avg,
+        }
+
+    def extract_measures(self, metric_query: str) -> tuple[list[Measure], ast.Query]:
+        """
+        Decomposes the metric query into its constituent aggregatable measures and
+        constructs a SQL query derived from those measures.
+        """
+        query_ast = parse(metric_query)
+        measures = []
+
+        for idx, func in enumerate(query_ast.find_all(ast.Function)):
+            handler = self.handlers.get(func.function())
+            if handler:
+                func_measures = handler(func, idx)
+                if func_measures:
+                    self.update_ast(func, func_measures)
+                measures.extend(func_measures)
+
+        return measures, query_ast
+
+    def _generic_additive_agg(self, func, idx) -> list[Measure]:
+        arg = func.args[0]
+        measure_name = "_".join(
+            [str(col) for col in arg.find_all(ast.Column)] + [func.name.name.lower()],
+        )
+        return [
+            Measure(
+                name=f"{measure_name}_{idx}",
+                expression=f"{func.quantifier} {arg}" if func.quantifier else str(arg),
+                aggregation=func.name.name.upper(),
+                rule=AggregationRule(
+                    type=Aggregability.FULL
+                    if func.quantifier != "DISTINCT"
+                    else Aggregability.LIMITED,
+                ),
+            ),
+        ]
+
+    def _avg(self, func, idx) -> list[Measure]:
+        arg = func.args[0]
+        measure_name = "_".join(
+            [str(col) for col in arg.find_all(ast.Column)]
+            + [dj_functions.Sum.__name__.lower()],
+        )
+        return [
+            Measure(
+                name=f"{measure_name}_{idx}",
+                expression=str(arg),
+                aggregation=dj_functions.Sum.__name__.upper(),
+                rule=AggregationRule(
+                    type=Aggregability.FULL
+                    if func.quantifier != "DISTINCT"
+                    else Aggregability.LIMITED,
+                ),
+            ),
+            Measure(
+                name="count",
+                expression="1",
+                aggregation=dj_functions.Count.__name__.upper(),
+                rule=AggregationRule(
+                    type=Aggregability.FULL
+                    if func.quantifier != "DISTINCT"
+                    else Aggregability.LIMITED,
+                ),
+            ),
+        ]
+
+    def update_ast(self, func, measures: list[Measure]):
+        """
+        Updates the query AST based on the measures derived from the function.
```

Comment:
How does this function make any updates to `self`? Is it by updating the `func` pointers?

Reply:
Yeah, you're right, it actually doesn't make any changes to `self` and just updates `func`. Let me refactor this method to be a `staticmethod` instead.